rest-client/src/main/java/io/registry/client/RegistryRestService.java

```diff
@@ -0,0 +1,487 @@
+import retrofit2.Retrofit;
+import retrofit2.converter.jackson.JacksonConverterFactory;
+
+import javax.enterprise.inject.Vetoed;
+import javax.ws.rs.Path;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.io.InputStream;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * @author Carles Arnal <carnalca@redhat.com>
+ */
+@Vetoed // not a bean
+@Path("__dummy_hack")
+public class RegistryRestService implements RegistryService {
+
+    private final Retrofit retrofit;
+    private ArtifactsService artifactsService;
+    private RulesService rulesService;
+    private SearchService searchService;
+    private IdsService idsService;
+
+    public RegistryRestService() {
+        // hack + client side only
+        retrofit = null;
+    }
+
+    protected RegistryRestService(String baseUrl) {
+
+        retrofit = new Retrofit.Builder()
+                .baseUrl(baseUrl)
+                .addConverterFactory(JacksonConverterFactory.create())
+                .build();
+
+        initServices(retrofit);
+    }
+
+    protected RegistryRestService(String baseUrl, OkHttpClient okHttpClient) {
+
+        retrofit = new Retrofit.Builder()
+                .client(okHttpClient)
+                .addConverterFactory(JacksonConverterFactory.create())
+                .baseUrl(baseUrl)
+                .build();
+
+        initServices(retrofit);
+    }
+
+    public static RegistryService create(String baseUrl) {
+        return new RegistryRestService(baseUrl);
+    }
+
+    public static RegistryService create(String baseUrl, OkHttpClient okHttpClient) {
+        return new RegistryRestService(baseUrl, okHttpClient);
+    }
+
+    private void initServices(Retrofit retrofit) {
+        artifactsService = retrofit.create(ArtifactsService.class);
+        rulesService = retrofit.create(RulesService.class);
+        idsService = retrofit.create(IdsService.class);
+        searchService = retrofit.create(SearchService.class);
+    }
+
+    @Override
+    public List<String> listArtifacts() {
+
+        final ResultCallback<List<String>> artifacts = new ResultCallback<>();
+
+        artifactsService.listArtifacts()
+                .enqueue(artifacts);
+
+        return artifacts.getResult();
+    }
+
+    @Override
+    public CompletionStage<ArtifactMetaData> createArtifact(ArtifactType artifactType, String artifactId, IfExistsType ifExistsType, InputStream data) {
+
+        final ResultCallback<ArtifactMetaData> result = new ResultCallback<>();
+
+        artifactsService.createArtifact(artifactType, artifactId, ifExistsType, RequestBody.create(null, IoUtil.toBytes(data)))
+                .enqueue(result);
+
+        return CompletableFuture.completedFuture(result.getResult());
+    }
+
+    @Override
+    public Response getLatestArtifact(String artifactId) {
+
+        final ResultCallback<ResponseBody> resultCallback = new ResultCallback<>();
```

Comment:
Some util or common method for this repeating task (new callback, call, enqueue, result, perhaps parse) would be nice ...

Reply:
Agree, that's in my todo list.